integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/lib"
+	"github.com/gravitational/teleport/lib/auth/testauthority"
+	"github.com/gravitational/teleport/lib/httplib/csrf"
+	"github.com/gravitational/teleport/lib/service"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/web"
+	"github.com/gravitational/teleport/lib/web/app"
+
+	"github.com/gravitational/trace"
+
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+// TestForward tests that requests within get forwarded to the target
+// application within a single cluster and through a trusted cluster.
+func TestForward(t *testing.T) {
+	// Create cluster, user, and credentials package.
+	pack := setup(t)
+
+	// Create user for tests.
+	pack.createUser(t)
+
+	// Create Web UI session.
+	webCookie, webToken := pack.createWebSession(t)
+
+	tests := []struct {
+		desc          string
+		inCookie      string
+		outStatusCode int
+		outMessage    string
+	}{
+		{
+			desc:          "root cluster, valid application session cookie, success",
+			inCookie:      pack.createAppSession(t, webCookie, webToken, pack.rootAppPublicAddr, pack.rootAppClusterName),
+			outStatusCode: http.StatusOK,
+			outMessage:    pack.rootMessage,
+		},
+		{
+			desc:          "leaf cluster, valid application session cookie, success",
+			inCookie:      pack.createAppSession(t, webCookie, webToken, pack.leafAppPublicAddr, pack.leafAppClusterName),
+			outStatusCode: http.StatusOK,
+			outMessage:    pack.leafMessage,
+		},
+		{
+			desc:          "invalid application session cookie, redirect to login",
+			inCookie:      "D25C463CD27861559CC6A0A6AE54818079809AA8731CB18037B4B37A80C4FC6C",
+			outStatusCode: http.StatusFound,
+			outMessage:    "",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			statusCode, message, err := pack.makeRequest(tt.inCookie)
+			require.NoError(t, err)
+			require.Equal(t, tt.outStatusCode, statusCode)
+			require.Contains(t, message, tt.outMessage)
+		})
+	}
+}
+
+// pack contains identity as well as initialized Teleport clusters and instances.
+type pack struct {
+	username string
+	password string
+
+	rootCluster        *TeleInstance
+	rootAppServer      *service.TeleportProcess
+	rootAppName        string
+	rootAppPublicAddr  string
+	rootAppClusterName string
+	rootMessage        string
+
+	leafCluster        *TeleInstance
+	leafAppServer      *service.TeleportProcess
+	leafAppName        string
+	leafAppPublicAddr  string
+	leafAppClusterName string
+	leafMessage        string
+}
+
+// setup configures all clusters and servers needed for a test.
+func setup(t *testing.T) *pack {
+	utils.InitLoggerForTests(testing.Verbose())
+
+	tr := utils.NewTracer(utils.ThisFunction()).Start()
+	defer tr.Stop()
+
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
```

Comment:
Can you clarify what this mode is for and whether it is intentional that it's limited for the lifetime of the setup function?